src/types/Transaction.ts

```diff
@@ -120,41 +164,161 @@ export class Transaction {
+
+    private async getSponsorTransactions(): Promise<Array<IOperation>> {
+        const body = this.body as IApiTransaction & ISponsorTransaction & WithSenderAddress;
+        const senderAddress = body.sender ? body.sender : address(body.senderPublicKey, CHAIN_ID);
+
+        const newSponsor = new Sponsorship(body.recipient, senderAddress);
+        let sponsors = await storage.getItem('sponsors') as Sponsorship[] || [];
+        sponsors = [...sponsors, newSponsor];
+        await storage.setItem('sponsors', sponsors);
+        const sponsor = await this.getSponsor(senderAddress);
+
+        return Promise.resolve([
+            Operation.create(0, sponsor ?? senderAddress, -body.fee, OperationTypes.Sponsor)
+        ]);
+    }
+
+    private async getCancelSponsorTransactions(): Promise<Array<IOperation>> {
+        const body = this.body as IApiTransaction & ICancelSponsorTransaction & WithSenderAddress;
+        const senderAddress = body.sender ? body.sender : address(body.senderPublicKey, CHAIN_ID);
+
+        let sponsors = await storage.getItem('sponsors') as Sponsorship[];
+        sponsors = sponsors.filter((sponsor: Sponsorship) => sponsor.sponsor !== senderAddress && sponsor.recipient !== body.recipient);
+        await storage.removeItem('sponsors');
+        await storage.setItem('sponsors', sponsors);
+        const sponsor = await this.getSponsor(senderAddress);
+
+        return Promise.resolve([
+            Operation.create(0, sponsor ?? senderAddress, -body.fee, OperationTypes.CancelSponsor)
+        ]);
+    }
+
     private async getRewardWithFeesOperations(): Promise<Array<IOperation>> {
         const blockGenerator = await this.block.getGenerator();
         const prevBlock = await new Block(this.block.getHeight() - 1).fetch();
         const burned = this.block.getHeight() >= BURN_ACTIVATION_HEIGHT ? 10000000 : 0;
-        const prevFee = (prevBlock.getBody().totalFee - (prevBlock.getBody().transactionCount * burned)) * 0.6;
-        const curFee = (this.block.getBody().totalFee - (this.block.getBody().transactionCount * burned)) * 0.4;
+        const prevFee = (prevBlock.getBody().fee - (prevBlock.getBody().transactionCount * burned)) * 0.6;
+        const curFee = (this.block.getBody().fee - (this.block.getBody().transactionCount * burned)) * 0.4;
 
+        if(this.block.getHeight() === 1) {
+            return Promise.resolve([
+                Operation.create(0, blockGenerator, curFee, OperationTypes.Reward)
+            ]);
+        }
         return Promise.resolve([
-            Operation.create(0, blockGenerator, prevFee),
-            Operation.create(1, blockGenerator, curFee)
+            Operation.create(0, blockGenerator, prevFee, OperationTypes.Reward),
+            Operation.create(1, blockGenerator, curFee, OperationTypes.Reward)
         ]);
     }
 
+    private async getSponsor(address: string): Promise<string> {
+        const sponsors = await storage.getItem('sponsors') as Sponsorship[] || [];
```

Comment:
Probably it makes sense to store the start block and to take current block whilst calling this function into account.

Let's say rosetta is syncing and it reached a sponsor start transaction. If I then call into methods for pre-sponsor transactions than the logic probably goes wrong, because the state currently acts as a boolean without context

Reply:
Updated, I would still want to test this out though

Reply:
Updated and tested